torch/distributed/_sharded_tensor/ops/linear.py

```diff
@@ -103,16 +107,24 @@ def sharded_linear(types, args, kwargs, pg):
 
     local_shard = weight.local_shards()[0].tensor
     local_shard_t = local_shard.t().contiguous()
+    local_shard.retain_grad()
```

Comment:
curious why we need to have `retain_grad()` call here? aren't the linear.weight already have `.grad`?

Reply:
This is something really tricky. If we don't set it, even if the `requre_grad` is true, the grad is `None`.

Reply:
Typically this needs to be done for intermediate values: https://stackoverflow.com/questions/45988168/why-does-autograd-not-produce-gradient-for-intermediate-variables. But `local_shard` is a leaf variable and not an intermediate variable in the autograd graph.

So I'm wondering why this needs to be done. Can you try printing out the autograd graph to see why we need `retain_grad`?

Reply:
This comment needs to be resolved too